radix-engine/src/kernel/track.rs

```diff
@@ -703,3 +717,288 @@ impl<'s> FinalizingTrack<'s> {
+                                .entry(vault_object_id)
+                                .or_default()
+                                .entry(resource_address)
+                                .or_insert(BalanceChange::Fungible(Decimal::ZERO))
+                                .fungible();
+                            existing.add_assign(delta);
+                        }
+                        BalanceChange::NonFungible { added, removed } => {
+                            let existing = direct_vault_updates
+                                .entry(vault_object_id)
+                                .or_default()
+                                .entry(resource_address)
+                                .or_insert(BalanceChange::NonFungible {
+                                    added: BTreeSet::new(),
+                                    removed: BTreeSet::new(),
+                                });
+                            existing.added_non_fungibles().extend(added);
+                            existing.removed_non_fungibles().extend(removed);
+                        }
+                    }
+                }
+            });
+
+        (balance_changes, direct_vault_updates)
+    }
+
+    fn traverse_state_updates(
+        &self,
+        balance_changes: &mut IndexMap<Address, IndexMap<ResourceAddress, BalanceChange>>,
+        accounted_vaults: &mut IndexSet<RENodeId>,
+        root: &Address,
+        current_node: &RENodeId,
+    ) -> () {
+        if let Some(modules) = self.indexed_state_updates.get(current_node) {
+            if self.is_vault(current_node) {
+                accounted_vaults.insert(current_node.clone());
+
+                if let Some((resource_address, balance_change)) =
+                    self.calculate_vault_balance_change(current_node)
+                {
+                    match balance_change {
+                        BalanceChange::Fungible(delta) => {
+                            let existing = balance_changes
+                                .entry(*root)
+                                .or_default()
+                                .entry(resource_address)
+                                .or_insert(BalanceChange::Fungible(Decimal::ZERO))
+                                .fungible();
+                            existing.add_assign(delta);
+                        }
+                        BalanceChange::NonFungible { added, removed } => {
+                            let existing = balance_changes
+                                .entry(*root)
+                                .or_default()
+                                .entry(resource_address)
+                                .or_insert(BalanceChange::NonFungible {
+                                    added: BTreeSet::new(),
+                                    removed: BTreeSet::new(),
+                                });
+                            existing.added_non_fungibles().extend(added);
```

Comment:
So I guess if someone moved from a non fungible vault to another under the same global component it would appear in both added and removed? makes sense

Reply:
Actually need some inputs on this - do you think we should prune the balance changes (for non-fungible, remove common entries among `added` and `removed`; for fungible, remove zero deltas)?

Reply:
I think this is fine as is :).

My input would be that in future IMO we should actually compute / store balance deltas at a per-vault level (which would avoid this problem), and let the consumers aggregate to the level they care about via the `parent_map` (see my other comment).